### PR TITLE
executor: Show placement information in SHOW TABLE x REGIONS

### LINKED ISSUE #32458: Show placement information in SHOW TABLE x REGIONS
Placement rules are by design asyncronously scheduled. The state of scheduling can be observed in the statement [`SHOW PLACEMENT FOR`.](https://docs.pingcap.com/tidb/stable/sql-statement-show-placement-for#examples)

This is a feature request to also include current placement information in the statement `SHOW TABLE x REGIONS`. The motivation for doing this is as follows:

- It allows observability of the current scheduling state (show placement for is too binary). Consider this as similar to how we can observe the progress of DDL
- It allows for verification that the rules are in fact followed correctly. We can incorporate this into our own scenario testing, but also for users that have compliance or HA requirements they might script against it to verify correctness.

The current output of `SHOW TABLE x REGIONS` is as follows:

```sql
tidb> show table t1 regions;
+-----------+---------------+---------------+-----------+-----------------+-------+------------+---------------+------------+----------------------+------------------+
| REGION_ID | START_KEY     | END_KEY       | LEADER_ID | LEADER_STORE_ID | PEERS | SCATTERING | WRITTEN_BYTES | READ_BYTES | APPROXIMATE_SIZE(MB) | APPROXIMATE_KEYS |
+-----------+---------------+---------------+-----------+-----------------+-------+------------+---------------+------------+----------------------+------------------+
|         4 |               | t_63_r_89565  |         5 |               1 | 5     |          0 |          4337 |      80007 |                  107 |           367623 |
|         6 | t_63_r_89565  | t_63_r_179845 |         7 |               1 | 7     |          0 |          1323 |   12187800 |                   91 |                0 |
|         8 | t_63_r_179845 | t_63_r_270126 |         9 |               1 | 9     |          0 |          1323 |   12187935 |                  102 |           115360 |
|        10 | t_63_r_270126 | t_63_r_360407 |        11 |               1 | 11    |          0 |            27 |   12187935 |                   91 |            63475 |
|        12 | t_63_r_360407 | t_63_r_450687 |        13 |               1 | 13    |          0 |          1323 |   12187800 |                  122 |           566573 |
|        14 | t_63_r_450687 | t_63_r_540968 |        15 |               1 | 15    |          0 |          1323 |    8125290 |                   95 |            57600 |
|        16 | t_63_r_540968 | t_63_r_631248 |        17 |               1 | 17    |          0 |          1323 |          0 |                  109 |           207822 |
|         2 | t_63_r_631248 |               |         3 |               1 | 3     |          0 |          1498 |          0 |                  103 |           104284 |
+-----------+---------------+---------------+-----------+-----------------+-------+------------+---------------+------------+----------------------+------------------+
8 rows in set (0.01 sec)
```

I propose that we add 2 columns:
- **Scheduling Constraints**: a list of recorded constraints for the region. It can be json formatted.
- **Scheduling State:** the same output as from `SHOW PLACEMENT FOR`.

### PR DESCRIPTION
**Pull request by johnhaxx7**
### What problem does this PR solve?

Issue Number: close #32458

Problem Summary:

Add `SCHEDULING_CONSTRAINTS` and `SCHEDULING_STATE` to `show table * regions`, not for `show table * index * regions`.

How to get the data is same as `show placement`:
`SCHEDULING_CONSTRAINTS` is from `information_schema`.
`SCHEDULING_STATE` is from pd api.

Output:
* For un-partitioned table, it's same as `show placement for table *`. 
* For partitioned table, result is fetched for each partition.

Sample output:
```
tidb> show table t3 regions;
+-----------+-----------+---------+-----------+-----------------+-------+------------+---------------+------------+----------------------+------------------+-------------------------------------------------------------------------------------------------------------------+------------------+
| REGION_ID | START_KEY | END_KEY | LEADER_ID | LEADER_STORE_ID | PEERS | SCATTERING | WRITTEN_BYTES | READ_BYTES | APPROXIMATE_SIZE(MB) | APPROXIMATE_KEYS | SCHEDULING_CONSTRAINTS                                                                                            | SCHEDULING_STATE |
+-----------+-----------+---------+-----------+-----------------+-------+------------+---------------+------------+----------------------+------------------+-------------------------------------------------------------------------------------------------------------------+------------------+
|        84 | t_95_     | t_96_   |        85 |               1 | 85    |          0 |             0 |          0 |                    0 |                0 | LEADER_CONSTRAINTS="[+region=us-east-1]" FOLLOWERS=3 FOLLOWER_CONSTRAINTS="[+region=us-east-2]"                   | PENDING          |
|        86 | t_96_     | t_97_   |        87 |               1 | 87    |          0 |             0 |          0 |                    0 |                0 | CONSTRAINTS="[+disk=ssd]" VOTERS=5 VOTER_CONSTRAINTS="[+region=bj]" LEARNERS=3 LEARNER_CONSTRAINTS="[+region=sh]" | PENDING          |
|        88 | t_97_     |         |        89 |               1 | 89    |          0 |             0 |          0 |                    0 |                0 | PRIMARY_REGION="cn-east-1" REGIONS="cn-east-1,cn-east-2" SCHEDULE="EVEN"                                          | PENDING          |
+-----------+-----------+---------+-----------+-----------------+-------+------------+---------------+------------+----------------------+------------------+-------------------------------------------------------------------------------------------------------------------+------------------+
```

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

```
create placement policy pa1 PRIMARY_REGION="cn-east-1" REGIONS="cn-east-1,cn-east-2"SCHEDULE="EVEN";
create placement policy pa2 LEADER_CONSTRAINTS="[+region=us-east-1]" FOLLOWER_CONSTRAINTS="[+region=us-east-2]" FOLLOWERS=3;
create placement policy pb1 VOTER_CONSTRAINTS="[+region=bj]" LEARNER_CONSTRAINTS="[+region=sh]" CONSTRAINTS="[+disk=ssd]"VOTERS=5 LEARNERS=3;
CREATE TABLE t3 (id INT) placement policy pa1 PARTITION BY RANGE (id) (PARTITION p0 VALUES LESS THAN (100) placement policy pa2,PARTITION p1 VALUES LESS THAN (1000) placement policy pb1,PARTITION p2 VALUES LESS THAN (10000));
```

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [x] Contains syntax changes - no sql syntax change but the output format changed, need to update public doc
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Support `SCHEDULING_CONSTRAINTS` and `SCHEDULING_STATE` 2 new columns for `show table regions` command.
```

### CODE REVIEW

File: executor/show.go
Comment (lcwangchao) [reply executor/show.go:1862]: And it's better to move those new added code to a new method to make the code easier to read.

File: executor/show.go
Comment (xhebox) [reply executor/show.go:1864]: The problem is that placement rules has been released. Our model may have been persisted. Thus if we want to add one, it may not be zero.

I think we could try to initialize every table/partition. Or ensure by an extra `map[xxx]bool`?

**Comment by lcwangchao** [comment]
Please fill the release note.

File: executor/show.go
Comment (lcwangchao) [reply executor/show.go:1864]: > The problem is that placement rules has been released. Our model may have been persisted. Thus if we want to add one, it may not be zero.
> 
> I think we could try to initialize every table/partition. Or ensure by an extra `map[xxx]bool`?

So some where stored the `PlacementScheduleState` as a persisted value?

Comment (xhebox) [reply executor/show.go:1864]: After some digging, I think no struct are referfing `PlacementScheduleState`. Thus we could add a new enumeration.

File: executor/show.go
Comment (johnhaxx7) [reply executor/show.go]: thanks for letting me know! I didn't aware of the split usage, added support and test for that case, PTAL

File: executor/split_table_test.go
Comment (johnhaxx7) [reply executor/split_table_test.go:648]: Updated, thanks!

File: executor/split.go
Comment (johnhaxx7) [reply executor/split.go]: Updated, PTAL and thanks for the info!

File: executor/show.go
Comment (johnhaxx7) [reply executor/show.go]: Updated, PTAL and thanks for the feedback!

File: executor/show.go
Comment (johnhaxx7) [reply executor/show.go]: Yep it looks suspicious indeed. 
Actually pending state is from line 60, because etcdCli is nil. Otherwise it'll fetch from PD. https://github.com/pingcap/tidb/blob/06737ecd29b0ae633d9ac89b0660e7a43939c88d/domain/infosync/region.go#L53-L69
Currently it reads from PD, @xhebox could you please explain a little bit more about "add a new enumeration"?

Comment (xhebox) [reply executor/show.go]: OK. The thing is that, `PlacementScheduleStatePending == 0`. So \ `var scheduleState map[int64]infosync.PlacementScheduleState` will get `scheduleState[xx] == PlacementScheduleStatePending` by default for any `xx` that is not fetched from PD.

> Because PlacementScheduleState.Pending is 0, so get value from an nil scheduleState will always return PlacementScheduleState.Pending , the test cases still passed even though the code is not right.

Comment (johnhaxx7) [reply executor/show.go]: Hi @xhebox, thanks for the reply! I think what you mentioned is's a pervasive issue that before this change. A typical usage of `fetchScheduleState` and `scheduleState` is:
https://github.com/pingcap/tidb/blob/2e1cff73f3bf7ee18d498a04138d39d70bd85a6d/executor/show_placement.go#L404-L415
In `fetchAllTablePlacements` function, `fetchScheduleState ` is called several times to fetch state from PD. `if block` at line 405 servers as a cache so that we don't need to ping PD if there's a match. But for key not found in `scheduleState`, `ok` will be false then it will go to line 410, call `GetReplicationState` to fetch from PD. I verified that in unit test, the pending state is returned from line 60 in `GetReplicationState ` function, not from the map. Please take a second look.

One more thing is we return pending state as a default state, even when there's something wrong with the system, which makes it error prone. But on a second thought, there's a `in progress` state, so call it pending still make sense. Or we can add another state to indicate if there's any system error cause the scheduling issue. But again, this is how states works before this change, if we think it's better to add another state, I think maybe I should start another PR, since there's many other places needs to update.

Comment (xhebox) [reply executor/show.go]: > But again, this is how states works before this change, if we think it's better to add another state, I think maybe I should start another PR, since there's many other places needs to update.

That makes sense. I agreed to leave it to the next PR, @lcwangchao , PTAL

Comment (johnhaxx7) [reply executor/show.go]: Thanks! Created a new issue here: https://github.com/pingcap/tidb/issues/35314, please feel free to let me know if we want to update the state

File: executor/split_table_test.go
Comment (lcwangchao) [inline executor/split_table_test.go:704]: Add some tests for `show table index regions` ?

File: executor/split.go
Comment (johnhaxx7) [inline executor/split.go:793]: Hi @lcwangchao, I agree with what you mentioned of avoiding to add `showTableRegionRowItem` to exist functions. But I need this physicalId in each regions to find the correct scheduling info for each region later. Can I add physicalID to regionMeta or if there's any other better way?

File: executor/split_table_test.go
Comment (johnhaxx7) [reply executor/split_table_test.go]: Done!

File: executor/show.go
Comment (johnhaxx7) [reply executor/show.go]: Done!

File: executor/split_table_test.go
Comment (johnhaxx7) [reply executor/split_table_test.go]: Done!

File: executor/split.go
Comment (lcwangchao) [reply executor/split.go:793]: I think it's reasonable to just add `physicalID` to `regionMeta` and the `getRegionMeta` should fill the right value for it.  Good point

File: executor/show.go
Comment (johnhaxx7) [reply executor/show.go]: Done!

File: executor/split.go
Comment (johnhaxx7) [reply executor/split.go]: Done!

File: executor/split.go
Comment (johnhaxx7) [reply executor/split.go]: Thanks, done!

**Comment by johnhaxx7** [comment]
Hi @lcwangchao, I resolve all the comments, can you please take a final look when you get a chance? 
Seems like `TestUpdateClusterIndex – executor` this test is blocking lots of prs, https://ci.pingcap.net/blue/organizations/jenkins/tidb_ghpr_unit_test/detail/tidb_ghpr_unit_test/53647/tests, I'm trying to rerun it and see..

File: planner/core/planbuilder.go
Comment (xhebox) [inline planner/core/planbuilder.go:2785]: Should this be documented in the RFC, @lcwangchao ?

File: executor/show.go
Comment (xhebox) [inline executor/show.go:1929]: What about an empty `schedulingState` if `constraintStr` is also empty?

File: executor/splittest/split_table_test.go
Comment (xhebox) [inline executor/splittest/split_table_test.go:281]: Why not explicitly `infosync.PlacementScheduleState.PENDING.String()`? Or maybe just "PENDING"?

**Review by xhebox — COMMENTED** [review]
LGTM Otherwise

File: executor/show.go
Comment (johnhaxx7) [reply executor/show.go]: Nice catch! Updated!

File: executor/splittest/split_table_test.go
Comment (johnhaxx7) [reply executor/splittest/split_table_test.go]: Hmm `infosync.PlacementScheduleState.PENDING.String() ` doesn't work and I think to reference the property instead of `PENDING` string is a better practice. So when someone update the property value, it won't break.

Comment (xhebox) [reply executor/splittest/split_table_test.go]: `require.Equal(t, infosync.PlacementScheduleStatePending.String(), rows[0][12])`

This does work.

File: executor/splittest/split_table_test.go
Comment (xhebox) [inline executor/splittest/split_table_test.go:299]: Assert the empty schedule state, too

**Review by lcwangchao — APPROVED** [review]

**Comment by johnhaxx7** [comment]
Hi @lcwangchao, can you please help me merge it when you get a chance? It keeps asking me to merge with master and when it finished there are some flaky tests blocking it. I'm in a different timezone, that will be great if you can help me merge it

File: executor/splittest/split_table_test.go
Comment (johnhaxx7) [reply executor/splittest/split_table_test.go]: seems like the code piece here is stale, it's in line 280-281, closing here, thanks!

**Review by xhebox — APPROVED** [review]

**Comment by xhebox** [comment]
> Hi @lcwangchao, can you please help me merge it when you get a chance? It keeps asking me to merge with master and when it finished there are some flaky tests blocking it. I'm in a different timezone, that will be great if you can help me merge it

merging needs 2 LGTM.

/merge

**Pull request merged**
